File: viddhana/packages/sdks/python/src/viddhana/vault.py
"""Vault module for portfolio and vault management."""

from typing import Optional, Dict, Any, List, TYPE_CHECKING
from web3 import Web3
import httpx

from .types import (
    Portfolio,
    Asset,
    PortfolioMetrics,
    VaultInfo,
    TransactionReceipt,
)

if TYPE_CHECKING:
    from eth_account.signers.local import LocalAccount


# Contract ABIs (simplified for SDK)
VAULT_MANAGER_ABI = [
    {
        "name": "createVault",
        "type": "function",
        "inputs": [{"name": "strategy", "type": "uint8"}],
        "outputs": [{"name": "vaultId", "type": "bytes32"}],
    },
    {
        "name": "deposit",
        "type": "function",
        "inputs": [
            {"name": "asset", "type": "address"},
            {"name": "amount", "type": "uint256"},
        ],
        "outputs": [],
    },
    {
        "name": "withdraw",
        "type": "function",
        "inputs": [
            {"name": "asset", "type": "address"},
            {"name": "amount", "type": "uint256"},
        ],
        "outputs": [],
    },
]

ERC20_ABI = [
    {
        "name": "approve",
        "type": "function",
        "inputs": [
            {"name": "spender", "type": "address"},
            {"name": "amount", "type": "uint256"},
        ],
        "outputs": [{"name": "", "type": "bool"}],
    },
    {
        "name": "balanceOf",
        "type": "function",
        "inputs": [{"name": "account", "type": "address"}],
        "outputs": [{"name": "", "type": "uint256"}],
    },
]


class VaultModule:
    """Module for vault and portfolio management."""

    # Contract addresses (configurable per network)
    VAULT_MANAGER_ADDRESS = "0x0000000000000000000000000000000000001001"

    def __init__(
        self,
        w3: Web3,
        account: Optional["LocalAccount"] = None,
        api_url: Optional[str] = None,
    ):
        """
        Initialize VaultModule.

        Args:
            w3: Web3 instance connected to Atlas chain.
            account: Optional account for signing transactions.
            api_url: API URL for portfolio queries.
        """
        self._w3 = w3
        self._account = account
        self._api_url = api_url
        self._http_client = httpx.Client(timeout=30.0)

    def _get_vault_manager(self):
        """Get VaultManager contract instance."""
        return self._w3.eth.contract(
            address=Web3.to_checksum_address(self.VAULT_MANAGER_ADDRESS),
            abi=VAULT_MANAGER_ABI,
        )

    def _sign_and_send(self, tx: Dict[str, Any]) -> TransactionReceipt:
        """Sign and send a transaction."""
        if not self._account:
            raise ValueError("Account required for transactions")

        # Build transaction
        tx["from"] = self._account.address
        tx["nonce"] = self._w3.eth.get_transaction_count(self._account.address)
        tx["gas"] = self._w3.eth.estimate_gas(tx)
        tx["gasPrice"] = self._w3.eth.gas_price

        # Sign and send
        signed = self._account.sign_transaction(tx)
        tx_hash = self._w3.eth.send_raw_transaction(signed.raw_transaction)

        # Wait for receipt
        receipt = self._w3.eth.wait_for_transaction_receipt(tx_hash)

        return TransactionReceipt(
            tx_hash=receipt["transactionHash"].hex(),
            block_number=receipt["blockNumber"],
            status=receipt["status"] == 1,
            gas_used=receipt["gasUsed"],
            effective_gas_price=receipt.get("effectiveGasPrice", 0),
            logs=[dict(log) for log in receipt["logs"]],
        )

    def create_vault(self, strategy: int = 0) -> TransactionReceipt:
        """
        Create a new vault.

        Args:
            strategy: Vault strategy type (0=conservative, 1=balanced, 2=aggressive).

        Returns:
            Transaction receipt.

        Raises:
            ValueError: If no account is configured.
        """
        if not self._account:
            raise ValueError("Account required to create vault")

        vault_manager = self._get_vault_manager()
        tx = vault_manager.functions.createVault(strategy).build_transaction({
            "chainId": self._w3.eth.chain_id,
        })

        return self._sign_and_send(tx)

    def deposit(self, asset: str, amount: int) -> TransactionReceipt:
        """
        Deposit assets into the vault.

        Args:
            asset: Token contract address.
            amount: Amount to deposit (in token's smallest unit).

        Returns:
            Transaction receipt.

        Raises:
            ValueError: If no account is configured.
        """
        if not self._account:
            raise ValueError("Account required for deposit")

        asset_address = Web3.to_checksum_address(asset)

        # First approve the vault manager to spend tokens
        token = self._w3.eth.contract(address=asset_address, abi=ERC20_ABI)
        approve_tx = token.functions.approve(
            Web3.to_checksum_address(self.VAULT_MANAGER_ADDRESS),
            amount,
        ).build_transaction({
            "chainId": self._w3.eth.chain_id,
        })

        self._sign_and_send(approve_tx)

        # Now deposit
        vault_manager = self._get_vault_manager()
        deposit_tx = vault_manager.functions.deposit(
            asset_address, amount
        ).build_transaction({
            "chainId": self._w3.eth.chain_id,
        })

        return self._sign_and_send(deposit_tx)

    def withdraw(self, asset: str, amount: int) -> TransactionReceipt:
        """
        Withdraw assets from the vault.

        Args:
            asset: Token contract address.
            amount: Amount to withdraw (in token's smallest unit).

        Returns:
            Transaction receipt.

        Raises:
            ValueError: If no account is configured.
        """
        if not self._account:
            raise ValueError("Account required for withdrawal")

        asset_address = Web3.to_checksum_address(asset)
        vault_manager = self._get_vault_manager()

        tx = vault_manager.functions.withdraw(
            asset_address, amount
        ).build_transaction({
            "chainId": self._w3.eth.chain_id,
        })

        return self._sign_and_send(tx)

    def get_portfolio(self, address: Optional[str] = None) -> Portfolio:
        """
        Get portfolio data for an address.

        Args:
            address: Address to query (defaults to connected account).

        Returns:
            Portfolio data.

        Raises:
            ValueError: If no address provided and no account configured.
        """
        if address is None:
            if not self._account:
                raise ValueError("Address required or account must be configured")
            address = self._account.address

        # Use JSON-RPC custom method
        result = self._w3.provider.make_request(
            "vdh_getPortfolio", [address]
        )

        if "error" in result:
            raise ValueError(result["error"]["message"])

        data = result["result"]

        assets = [
            Asset(
                symbol=a["symbol"],
                address=a["address"],
                balance=a["balance"],
                value=a["value"],
                allocation=a["allocation"],
            )
            for a in data.get("assets", [])
        ]

        metrics_data = data.get("metrics", {})
        metrics = PortfolioMetrics(
            volatility_30d=metrics_data.get("volatility30d", 0.0),
            sharpe_ratio=metrics_data.get("sharpeRatio", 0.0),
            unrealized_pnl=metrics_data.get("unrealizedPnL", 0.0),
            unrealized_pnl_percent=metrics_data.get("unrealizedPnLPercent", 0.0),
        )

        return Portfolio(
            address=data["address"],
            total_value=data["totalValue"],
            currency=data["currency"],
            assets=assets,
            metrics=metrics,
            last_updated=data["lastUpdated"],
        )

    def get_vault_info(self, vault_id: str) -> VaultInfo:
        """
        Get vault information.

        Args:
            vault_id: Vault identifier.

        Returns:
            VaultInfo with vault details.
        """
        result = self._w3.provider.make_request(
            "vdh_getVaultInfo", [vault_id]
        )

        if "error" in result:
            raise ValueError(result["error"]["message"])

        data = result["result"]

        assets = [
            Asset(
                symbol=a["symbol"],
                address=a["address"],
                balance=a["balance"],
                value=a["value"],
                allocation=a["allocation"],
            )
            for a in data.get("assets", [])
        ]

        return VaultInfo(
            vault_id=data["vaultId"],
            owner=data["owner"],
            assets=assets,
            total_value=data["totalValue"],
            strategy=data["strategy"],
            created_at=data["createdAt"],
        )

    def close(self):
        """Close HTTP client."""
        self._http_client.close()

File: viddhana/packages/sdks/javascript/src/modules/vault.ts
/**
 * Vault Module - Portfolio and vault operations
 */

import { ethers } from 'ethers';
import type {
  PortfolioData,
  VaultInfo,
  RebalanceHistory,
  RebalanceAction,
} from '../types';

// Contract ABIs (minimal interfaces)
const ERC20_ABI = [
  'function approve(address spender, uint256 amount) returns (bool)',
  'function balanceOf(address owner) view returns (uint256)',
  'function allowance(address owner, address spender) view returns (uint256)',
];

const VAULT_MANAGER_ABI = [
  'function deposit(address asset, uint256 amount) returns (uint256)',
  'function withdraw(address asset, uint256 amount) returns (uint256)',
  'function createVault(string name, string symbol, address[] assets, uint256[] allocations) returns (address)',
  'function getVaultInfo(address vault) view returns (tuple(address, string, string, uint256, uint256, uint256, string))',
  'function getUserVaults(address user) view returns (address[])',
];

const POLICY_ENGINE_ABI = [
  'function executeManualRebalance(tuple(address asset, uint8 action, uint256 percentage, uint256 amount, uint256 valueUSD)[] actions) returns (bool)',
  'function setUserProfile(uint256 riskTolerance, uint256 timeToGoal, bool autoRebalance) returns (bool)',
  'function getUserProfile(address user) view returns (tuple(uint256, uint256, bool))',
];

// Contract addresses (to be configured per network)
const CONTRACT_ADDRESSES = {
  mainnet: {
    vaultManager: '0x0000000000000000000000000000000000000001',
    policyEngine: '0x0000000000000000000000000000000000000002',
  },
  testnet: {
    vaultManager: '0x0000000000000000000000000000000000000001',
    policyEngine: '0x0000000000000000000000000000000000000002',
  },
};

/**
 * Vault module for portfolio management and vault operations
 *
 * @example
 * ```typescript
 * // Get portfolio
 * const portfolio = await client.vault.getPortfolio('0x...');
 * console.log(`Total Value: $${portfolio.totalValue}`);
 *
 * // Deposit to vault
 * const tx = await client.vault.deposit('0xUSDC...', ethers.parseUnits('1000', 6));
 * await tx.wait();
 * ```
 */
export class Vault {
  private readonly provider: ethers.JsonRpcProvider;
  private readonly signer: ethers.Wallet | null;
  private readonly apiUrl: string;

  constructor(
    provider: ethers.JsonRpcProvider,
    signer: ethers.Wallet | null,
    apiUrl: string
  ) {
    this.provider = provider;
    this.signer = signer;
    this.apiUrl = apiUrl;
  }

  /**
   * Gets portfolio data for an address
   *
   * @param address - The address to get portfolio for
   * @returns Portfolio data including assets, values, and metrics
   */
  async getPortfolio(address: string): Promise<PortfolioData> {
    const result = await this.provider.send('vdh_getPortfolio', [address]);
    return result as PortfolioData;
  }

  /**
   * Gets rebalancing history for an address
   *
   * @param address - The address to get history for
   * @param options - Optional filter options
   * @returns Rebalance history with transaction details
   */
  async getRebalanceHistory(
    address: string,
    options?: {
      fromBlock?: number;
      toBlock?: number | 'latest';
      limit?: number;
    }
  ): Promise<RebalanceHistory> {
    const result = await this.provider.send('vdh_getRebalanceHistory', [
      address,
      options ?? {},
    ]);
    return result as RebalanceHistory;
  }

  /**
   * Creates a new vault
   *
   * @param name - Vault name
   * @param symbol - Vault token symbol
   * @param assets - Array of asset addresses
   * @param allocations - Array of allocation percentages (basis points, sum = 10000)
   * @returns Transaction response
   */
  async createVault(
    name: string,
    symbol: string,
    assets: string[],
    allocations: number[]
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const vaultManager = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vaultManager,
      VAULT_MANAGER_ABI,
      this.signer!
    );

    const tx = await vaultManager.createVault(name, symbol, assets, allocations);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Deposits assets into a vault
   *
   * @param asset - Asset token address
   * @param amount - Amount to deposit (in wei/smallest unit)
   * @returns Transaction response
   */
  async deposit(
    asset: string,
    amount: bigint
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    // First approve the token transfer
    const token = new ethers.Contract(asset, ERC20_ABI, this.signer!);
    const approveTx = await token.approve(
      CONTRACT_ADDRESSES.mainnet.vaultManager,
      amount
    );
    await approveTx.wait();

    // Then execute deposit
    const vaultManager = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vaultManager,
      VAULT_MANAGER_ABI,
      this.signer!
    );

    const tx = await vaultManager.deposit(asset, amount);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Withdraws assets from a vault
   *
   * @param asset - Asset token address
   * @param amount - Amount to withdraw (in wei/smallest unit)
   * @returns Transaction response
   */
  async withdraw(
    asset: string,
    amount: bigint
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const vaultManager = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vaultManager,
      VAULT_MANAGER_ABI,
      this.signer!
    );

    const tx = await vaultManager.withdraw(asset, amount);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Gets vault information
   *
   * @param vaultAddress - Vault contract address
   * @returns Vault information
   */
  async getVaultInfo(vaultAddress: string): Promise<VaultInfo> {
    const vaultManager = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vaultManager,
      VAULT_MANAGER_ABI,
      this.provider
    );

    const info = await vaultManager.getVaultInfo(vaultAddress);
    return {
      address: info[0] as string,
      name: info[1] as string,
      symbol: info[2] as string,
      totalAssets: info[3].toString(),
      totalShares: info[4].toString(),
      apy: Number(info[5]) / 100,
      strategy: info[6] as string,
    };
  }

  /**
   * Gets all vaults for a user
   *
   * @param userAddress - User's address
   * @returns Array of vault addresses
   */
  async getUserVaults(userAddress: string): Promise<string[]> {
    const vaultManager = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.vaultManager,
      VAULT_MANAGER_ABI,
      this.provider
    );

    return vaultManager.getUserVaults(userAddress) as Promise<string[]>;
  }

  /**
   * Executes a manual rebalance
   *
   * @param actions - Array of rebalance actions
   * @returns Transaction response
   */
  async executeRebalance(
    actions: RebalanceAction[]
  ): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const policyEngine = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.policyEngine,
      POLICY_ENGINE_ABI,
      this.signer!
    );

    const formattedActions = actions.map((action) => ({
      asset: action.asset,
      action: action.action === 'BUY' ? 0 : 1,
      percentage: action.percentage,
      amount: ethers.parseEther(action.amount),
      valueUSD: ethers.parseUnits(action.valueUSD, 6),
    }));

    const tx = await policyEngine.executeManualRebalance(formattedActions);
    return tx as ethers.TransactionResponse;
  }

  /**
   * Sets user profile for auto-rebalancing
   *
   * @param params - Profile parameters
   * @returns Transaction response
   */
  async setProfile(params: {
    riskTolerance: number; // 0-10000 basis points
    timeToGoal: number; // months
    autoRebalance: boolean;
  }): Promise<ethers.TransactionResponse> {
    this.requireSigner();

    const policyEngine = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.policyEngine,
      POLICY_ENGINE_ABI,
      this.signer!
    );

    const tx = await policyEngine.setUserProfile(
      params.riskTolerance,
      params.timeToGoal,
      params.autoRebalance
    );
    return tx as ethers.TransactionResponse;
  }

  /**
   * Gets user profile settings
   *
   * @param address - User address
   * @returns User profile
   */
  async getProfile(address: string): Promise<{
    riskTolerance: number;
    timeToGoal: number;
    autoRebalance: boolean;
  }> {
    const policyEngine = new ethers.Contract(
      CONTRACT_ADDRESSES.mainnet.policyEngine,
      POLICY_ENGINE_ABI,
      this.provider
    );

    const profile = await policyEngine.getUserProfile(address);
    return {
      riskTolerance: Number(profile[0]),
      timeToGoal: Number(profile[1]),
      autoRebalance: profile[2] as boolean,
    };
  }

  /**
   * Gets AI optimization recommendation
   *
   * @param address - User address
   * @returns Optimization recommendation
   */
  async getOptimizationRecommendation(address: string): Promise<{
    actions: RebalanceAction[];
    confidence: number;
    reason: string;
  }> {
    const portfolio = await this.getPortfolio(address);

    const response = await fetch(`${this.apiUrl}/v1/optimize/portfolio`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        user_id: address,
        portfolio,
      }),
    });

    if (!response.ok) {
      throw new Error(`API error: ${response.status}`);
    }

    return response.json() as Promise<{
      actions: RebalanceAction[];
      confidence: number;
      reason: string;
    }>;
  }

  /**
   * Checks token allowance
   *
   * @param tokenAddress - Token contract address
   * @param ownerAddress - Token owner address
   * @param spenderAddress - Spender address (defaults to vault manager)
   * @returns Allowance amount
   */
  async getAllowance(
    tokenAddress: string,
    ownerAddress: string,
    spenderAddress?: string
  ): Promise<bigint> {
    const token = new ethers.Contract(tokenAddress, ERC20_ABI, this.provider);
    return token.allowance(
      ownerAddress,
      spenderAddress ?? CONTRACT_ADDRESSES.mainnet.vaultManager
    ) as Promise<bigint>;
  }

  /**
   * Gets token balance
   *
   * @param tokenAddress - Token contract address
   * @param ownerAddress - Token owner address
   * @returns Token balance
   */
  async getTokenBalance(
    tokenAddress: string,
    ownerAddress: string
  ): Promise<bigint> {
    const token = new ethers.Contract(tokenAddress, ERC20_ABI, this.provider);
    return token.balanceOf(ownerAddress) as Promise<bigint>;
  }

  /**
   * Requires signer to be present
   * @private
   */
  private requireSigner(): void {
    if (!this.signer) {
      throw new Error('Signer required for this operation. Provide privateKey in config.');
    }
  }
}
